### ISSUE TITLE
Ability to adopt existing `.app`s as Cask artifacts

### ISSUE THREAD

**Issue by ThatsJustCheesy** [Homebrew/brew#14006]
### Provide a detailed description of the proposed feature

A flag to `brew install --cask` (say `--adopt`) that skips the artifact [move](https://github.com/Homebrew/brew/blob/8684f953f621db5f1a1eed0477f34b0ee5315ed1/Library/Homebrew/cask/artifact/moved.rb#L37-L67) / [link](https://github.com/Homebrew/brew/blob/978a452ea7ccdff1ed7377f88add3ff0513ba2b7/Library/Homebrew/cask/artifact/symlinked.rb#L48-L70) steps when the target already exists and has the same checksum as the source (or doesn't; see alternatives considered).

To illustrate, if "there is already an X at '/path/to/target'"

```
$ brew install textmate
==> Downloading https://github.com/textmate/textmate/releases/download/v2.0.23/T
Already downloaded: /Users/ian/Library/Caches/Homebrew/downloads/51ffb61879f6a076ae32e92cd9363820ee0d909e9b03015448bc0d0728970273--TextMate_2.0.23.tbz
==> Installing Cask textmate
==> Purging files for version 2.0.23 of Cask textmate
Error: It seems there is already an App at '/Applications/TextMate.app'.
$
```

users currently have two options:

- Use the artifact that's already installed, without it being managed by Homebrew
- Use `--force` to overwrite the installed binary, potentially changing it to some different version

This feature would add a third option:

- Use `--adopt` (or whatever) to have Homebrew treat the existing artifact as if it was installed from a cask

### What is the motivation for the feature?

This would be a useful option to migrate an existing system to use casks, reaping the associated benefits:

- Ability use to `brew upgrade` for apps that don't auto-update
- Ability to use `brew uninstall` for easy or mass uninstallation
- Ability to use `brew uninstall --zap` to nuke all files
- Inclusion in `brew bundle dump`
- etc.

without risking installing a different artifact by accident, with `--force`.

&lt;selfish reason, ignore if uninterested&gt;

My [Homebrew GUI client](https://getovert.app/about/) would also benefit, as there would be less confusion about apps installed from the web showing as "Available" but then failing to install. If the failure were instead an `--adopt`, all of the benefits listed above involving the command line would apply inside the GUI, too.

&lt;/selfish reason, ignore if uninterested&gt;

### How will the feature be relevant to at least 90% of Homebrew users?

Every Homebrew user has installed apps manually from the web before.

It's highly likely that many of these apps exists as casks, but were not installed that way. Once users learn about casks, I believe many would be willing to use them manage all of their apps. However, there is currently no easy way to migrate without passing scary `--force` flags or manually trashing existing `.app`s.

### What alternatives to the feature have been considered?

- Add `--adopt`, but don't have it verify the checksum.
  - This would allow the "installed version" reported by `brew info` to not match the `.app`'s real version; **however, cask auto-updates already do this**, so perhaps this is a moot point.
  - If the above point is not seen as a big deal, I would actually prefer this behaviour.
- &lt;bikeshedding&gt; Call it `--adopt-existing` or something else. &lt;/bikeshedding&gt;

---

I'd be willing to attempt a PR for this if sentiment is positive.

**Comment by MikeMcQuaid** [comment]
> I'd be willing to attempt a PR for this if sentiment is positive.

Yeh, I'd be game for this as long as it verified the nested file contents and checksums match.

My only fear is that this will end up being pretty slow if you're gonna do a comparison between two big `.app`. May be better with a recursive (fail fast) `diff` rather than checksum validation on every file.

**Comment by ThatsJustCheesy** [comment]
I'm in the middle of implementing this, and have realized the combination of `--adopt` and `--force` is somewhat ambiguous. It could:

1. Ignore the result of the `diff`, and just pretend the source and target are the same
2. Overwrite the target, as if `--adopt` wasn't even given
3. Just raise an error, leaving a possible future extension point

Any thoughts? I personally like option 1 again, but I can also think of reasonable arguments for options 2 and 3.

**Comment by Rylan12** [comment]
Personally, I like option 1. Adding `--adopt` shows that the user wants something different from just `--force` so I think it's fine to have it function differently in that case.

**Comment by apainintheneck** [comment]
Closed by #14033. Thanks @ThatsJustCheesy!
